### DISCUSSION TITLE
emqx.conf in a custom directory for v4.x

### DISCUSSION THREAD

**Discussion by naikvidish** [emqx/emqx#8497] [Q&A] [answered]
Hi, 

I have a couple of question with respect to emqx.conf:

1. Is there a way to store and read the emqx.conf from a directory other than `etc`. How do I let EMQX know about it. 
2. If not, is there a way to add additional conf files or override files similar to the ones in v5?
3. Going over the config, I found the value `{{ platform_etc_dir }}` which is substituted by the actual value. When is this `{{ platform_etc_dir }}` value replaced by the actual value?  

Thanks in advance!

**Comment by zmstone** [comment, marked as answer]
1. Yes, but it has to be done manually, including the “vars” file which is generated from:
https://github.com/emqx/emqx/blob/cd8a4503f897f68be8eb341b69e123394cd3eaea/data/emqx_vars#L12

2. No. You’ll need to merge configs in one file
3. The substitution is done at compile time per build profile. If you want to build from source code, rebar.config.erl is where the values are defined

**Reply by naikvidish** [reply]
Thanks for the response!

**Reply by naikvidish** [reply]
Hi @zmstone,

Thanks once again for the suggestions. I tried them and EMQX broker was able to pick up the config from the custom directory.

I have a use case where I want to restart the broker from the code itself. I used the shutdown and reboot functions and they have worked for me. The issue I am facing is that after the restart, EMQX falls back to reading from the default `etc` directory and not the custom directory. 
Can you help me understand the cause behind it and is there a way to fix it?

**Reply by zmstone** [reply]
Could you be more specific about 

* which did you change for the etc path?
* which exact shutdown and restart functions used?

**Reply by naikvidish** [reply]
As you suggested, I substituted `runner_etc_dir` path [here](https://github.com/emqx/emqx/blob/v4.4.5/rebar.config.erl#L220) in the `rebar.config.erl` to the path to custom directory before building the broker. I also changed the `plugins.etc_dir` and `acl_file` configs to point to the custom directory. 
I used this [shutdown](https://github.com/emqx/emqx/blob/v4.4.5/src/emqx.erl#L226) and this [reboot](https://github.com/emqx/emqx/blob/v4.4.5/src/emqx.erl#L236)

**Reply by zmstone** [reply]
Thank you.
And how did you observe the fallback after reboot ?

BTW. The shutdown and reboot are exported for `ekka` to call after joining a cluster.
I don't think it's intended for other use cases.

**Reply by naikvidish** [reply]
I had 2 etc directories. One in the default directory and the other in the custom directory. When EMQX starts, it picks up the config from custom location and I can verify it since I changed the `listener.ssl.external` option to a different port.
Before restarting, I would make different changes to `emqx.conf` in both directories and I could see that the changes in the default etc directory were picked up. 

If I completely remove the default etc directory, EMQX still uses the custom directory for the initial start but it would fail to restart.
